pkg/cloudprovider/providers/openstack/openstack_volumes.go

```diff
@@ -327,10 +334,70 @@ func (os *OpenStack) GetDevicePath(volumeID string) string {
 		}
 	}
 
-	glog.Warningf("Failed to find device for the volumeID: %q\n", volumeID)
+	glog.V(4).Infof("Failed to find device for the volumeID: '%q' by serial ID.\n", volumeID)
+	return ""
+}
+
+func (os *OpenStack) GetDevicePathFromInstanceMetadata(volumeID string) string {
+	// Nova Hyper-V hosts cannot override disk SCSI IDs. In order to locate
+	// volumes, we're querying the metadata service. Note that the Hyper-V
+	// driver will include device metadata for untagged volumes as well.
+	//
+	// We're avoiding using cached metadata (or the configdrive),
+	// relying on the metadata service.
+	instanceMetadata, err := getMetadataFromMetadataService(
+		NewtonMetadataVersion)
+
+	if err != nil {
+		glog.V(4).Infof(
+			"Could not retrieve instance metadata. Error: %v.\n", err)
+		return ""
+	}
+
+	for _, device := range instanceMetadata.Devices {
+		if device.Type == "disk" && device.Serial == volumeID {
+			glog.V(4).Infof(
+				`Found disk metadata for volumeID '%s'.
+				 Bus: '%s', Address: '%s'`,
+				volumeID, device.Bus, device.Address)
+
+			diskPattern := fmt.Sprintf("/dev/disk/by-path/*-%s-%s",
+									   device.Bus, device.Address)
+			diskPaths, err := filepath.Glob(diskPattern)
+			if err != nil {
+				glog.Errorf(
+					`Could not retrieve disk path for volumeID: '%s'.
+					Error filepath.Glob(\"%s\"): %v\n`,
+					volumeID, diskPattern, err)
+				return ""
+			}
+
+			if len(diskPaths) == 1 {
```

Comment:
How about checking "0" and ">1"?
If "0", we can return "" soon.

Reply:
Good point, I'm updating the commit.

Reply:
Updated the commit, immediately returning if getting none or multiple disk paths.

While at it, I've also updated the log messages according to the k8s developer guide. Some of them were not formatted properly.

Reply:
Have you updated the commit?
But I can't find the codes which check "0".

Reply:
Yes, well, I don't explicitly check for '0': if there's exactly one disk path, I return it, otherwise I'm logging an error message and returning an empty string immediately: https://github.com/petrutlucian94/kubernetes/blob/fea1939cd7684169357fc13fd2ff4cd28e9d8eed/pkg/cloudprovider/providers/openstack/openstack_volumes.go#L375-L383

(in case the previous link becomes unavailable due to a new commit, here's the snippet: http://paste.openstack.org/raw/625887/)

Reply:
All right, thank you.